source/common/router/scoped_rds.cc

```diff
@@ -121,19 +121,87 @@ ScopedRdsConfigSubscription::ScopedRdsConfigSubscription(
     envoy::extensions::filters::network::http_connection_manager::v3::Rds& rds,
     Init::Manager& init_manager)
-    : parent_(parent), scope_name_(scope_name),
-      route_provider_(std::dynamic_pointer_cast<RdsRouteConfigProviderImpl>(
-          parent_.route_config_provider_manager_.createRdsRouteConfigProvider(
-              rds, parent_.factory_context_, parent_.stat_prefix_, init_manager))),
+    : parent_(parent), scope_name_(scope_name), on_demand_(false) {
+  initRdsConfigProvider(rds, init_manager);
+}
+
+// Constructor for RdsRouteConfigProviderHelper when scope is on demand.
+// Leave the RdsRouteConfigProvider uninitialized.
+ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::RdsRouteConfigProviderHelper(
+    ScopedRdsConfigSubscription& parent, std::string scope_name)
+    : parent_(parent), scope_name_(scope_name), on_demand_(true) {
+  parent_.stats_.on_demand_scopes_.inc();
+}
+
+void ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::addOnDemandUpdateCallback(
+    std::function<void()> callback) {
+  // If route table has been initialized, run the callback to continue in filter chain, otherwise
+  // cache it and wait for the route table to be initialized.
+  if (route_provider_ != nullptr && routeConfig() != std::make_shared<NullConfigImpl>()) {
+    callback();
+    return;
+  }
+  on_demand_update_callbacks_.push_back(callback);
+  // Initialize the rds provider if it has not been initialized.
+  maybeInitRdsConfigProvider();
+}
+
+void ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::runOnDemandUpdateCallback() {
+  for (auto& callback : on_demand_update_callbacks_) {
+    callback();
+  }
+  on_demand_update_callbacks_.clear();
+}
+
+void ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::initRdsConfigProvider(
+    envoy::extensions::filters::network::http_connection_manager::v3::Rds& rds,
+    Init::Manager& init_manager) {
+  route_provider_ = std::dynamic_pointer_cast<RdsRouteConfigProviderImpl>(
+      parent_.route_config_provider_manager_.createRdsRouteConfigProvider(
+          rds, parent_.factory_context_, parent_.stat_prefix_, init_manager));
+
+  rds_update_callback_handle_ = route_provider_->subscription().addUpdateCallback([this]() {
+    // Subscribe to RDS update.
+    parent_.onRdsConfigUpdate(scope_name_, route_provider_->subscription());
+  });
+  parent_.stats_.active_scopes_.inc();
+}
 
-      rds_update_callback_handle_(route_provider_->subscription().addUpdateCallback([this]() {
-        // Subscribe to RDS update.
-        parent_.onRdsConfigUpdate(scope_name_, route_provider_->subscription());
-      })) {}
+void ScopedRdsConfigSubscription::RdsRouteConfigProviderHelper::maybeInitRdsConfigProvider() {
+  // If the route provider have been initialized, return and wait for rds config update.
+  if (route_provider_) {
+    return;
+  }
+  // Create a init_manager to create a rds provider.
+  std::unique_ptr<Init::ManagerImpl> srds_init_mgr =
```

Comment:
Doesn't this need to live through the entire warming of the RDS provider?

Reply:
why? 
this init-mgr is a noop init-mgr and it only starts the rds subscription. the notification that "a routeConfiguration is back " is done in the ScopedRdsConfigSubscription::onRdsConfigUpdate.

Reply:
I'm thinking about transitive warming dependencies, but looking at the call graph, it seems we only get here for on-demand updates, so I agree, this is safe. Maybe add a comment explaining this.

Reply:
Sure, will add some comments.